src/execution/mod.rs

```diff
@@ -41,277 +38,485 @@ impl Pump {
-
-                shutdown_worker(worker);
-
-                Ok(())
+#[pymethods]
+impl WorkerCoro {
+    /// "Main" loop of this worker thread.
+    ///
+    /// 1. Pump `Pump`s (get input data from Python into Timely).
+    ///
+    /// 2. Dispose of empty `Pump`s which indicate there's no more
+    /// input data.
+    ///
+    /// 3. Step the worker to do whatever work it can. This will
+    /// execute operators.
+    ///
+    /// 4. Dispose of empty probes which indicates that all output has
+    /// been emitted from that caputre operator.
+    ///
+    /// 5. Determine if the worker should exit.
+    fn send(&mut self, py: Python, _value: Py<PyAny>) -> PyResult<()> {
+        let timely_worker = self.timely_worker.as_mut().ok_or(PyRuntimeError::new_err(
+            "Dataflow previously threw unhandled exception; can't be used again",
+        ))?;
+
+        for pump in self.input_pumps.iter_mut() {
+            pump.pump(py)?;
+        }
+        self.input_pumps = self
+            .input_pumps
+            .drain(..)
+            .filter(Pump::input_remains)
+            .collect();
+
+        // SAFETY: TWorker must outlive the catch_unwind boundary to
+        // allow async resume. But TWorker is implemented in Timely
+        // using some things that are not UnwindSafe. Thus after a
+        // panic, throw away the TWorker via assigning None so logic
+        // safety can be enforced by never being able to use the
+        // possibly-corrupt TWorker again.
+        std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+            py.allow_threads(|| {
+                timely_worker.step();
             })
-        })
-    });
-
-    match result {
-        Ok(Ok(ok)) => Ok(ok),
-        Ok(Err(build_err_str)) => Err(PyValueError::new_err(build_err_str)),
-        Err(panic_err) => {
-            let pyerr = if let Some(pyerr) = panic_err.downcast_ref::<PyErr>() {
+        }))
+        .map_err(|err| {
+            // Never allow calls to this possibly-corrupted by
+            // panic TWorker again.
+            self.timely_worker = None;
+
+            if let Some(pyerr) = err.downcast_ref::<PyErr>() {
                 pyerr.clone_ref(py)
             } else {
                 PyRuntimeError::new_err("Panic in Rust code")
-            };
-            Err(pyerr)
+            }
+        })?;
+
+        self.output_probes = self
+            .output_probes
+            .drain(..)
+            .filter(|probe| !probe.done())
+            .collect();
+
+        let should_stop = self.should_stop.call0(py)?.extract(py)?;
+        let work_complete = self.input_pumps.is_empty() && self.output_probes.is_empty();
+
+        if should_stop || work_complete {
+            self.close()?;
+            Err(PyStopIteration::new_err(()))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Run code on shutdown of worker.
+    fn close(&mut self) -> PyResult<()> {
```

Comment:
`close()` is described as "raising `GeneratorExit` in the code, and the result of that is shutting down. Hence, no exceptions here.